src/jit/hwintrinsicxarch.cpp

```diff
@@ -218,6 +228,39 @@ int Compiler::numArgsOfHWIntrinsic(NamedIntrinsic intrinsic)
     return hwIntrinsicInfoArray[intrinsic - NI_HW_INTRINSIC_START - 1].numArgs;
 }
 
+//------------------------------------------------------------------------
+// numArgsOfHWIntrinsic: get the number of arguments based on baseType
+//
+// Arguments:
+//    intrinsic -- id of the intrinsic function.
+//    baseType  -- var_types type of intrinsic
+//
+// Return Value:
+//     number of arguments
+//
+int Compiler::numArgsOfHWIntrinsic(NamedIntrinsic intrinsic, var_types baseType)
+{
+    assert(intrinsic != NI_Illegal);
+    assert(intrinsic > NI_HW_INTRINSIC_START && intrinsic < NI_HW_INTRINSIC_END);
+
+    int numArgs = hwIntrinsicInfoArray[intrinsic - NI_HW_INTRINSIC_START - 1].numArgs;
+    if (numArgs == -1)
+    {
+        switch (intrinsic)
+        {
+            case NI_SSE2_Shuffle:
+                assert(hwIntrinsicArgsInfoArray[0].intrinsicID == intrinsic);
+                assert((baseType >= TYP_BYTE) && (baseType <= TYP_DOUBLE));
+                return hwIntrinsicArgsInfoArray[0].numArgs[baseType - TYP_BYTE];
+
+            default:
+                break;
+        }
+    }
+
+    return numArgs;
+}
```

Comment:
Finally, this function could be somthing like:
```cpp
int Compiler::numArgsOfHWIntrinsic(NamedIntrinsic intrinsic, GenTreeHWIntrinsic* node)
{
    assert(intrinsic != NI_Illegal);
    assert(intrinsic > NI_HW_INTRINSIC_START && intrinsic < NI_HW_INTRINSIC_END);
    if ((flagsOfHWIntrinsic(intrinsic) & HW_Flag_UnfixedNumArgs) == 0)
    {
        return hwIntrinsicInfoArray[intrinsic - NI_HW_INTRINSIC_START - 1].numArgs;
    }

    assert(hwIntrinsicInfoArray[intrinsic - NI_HW_INTRINSIC_START - 1].numArgs == -1);

    GenTree* op1 = node->gtGetOp1();
    GenTree* op2 = node->gtGetOp2();
    if (op1 != nullptr)
    {
        if (op2 != nullptr)
        {
            return 2;
        }
        else if (op1->OperIsList())
        {
            assert(op1->AsArgList()->Rest()->Rest()->Current() != nullptr);
            assert(op1->AsArgList()->Rest()->Rest()->Rest() == nullptr);
            return 3;
        }
        else
        {
            return 1;
        }
    }
    else
    {
        return 0;
    }
}
```

Reply:
Working on your feedback

Reply:
After nalyzing your proposal to replace existing implementation for retrieving numArgs with your proposal I have found the following disadvantages:

1. It introduces additional flag `HW_Flag_UnfixedNumArgs` which is not necessary since table was designed in way to provide information if given intrinsic has variable number of arguments in signature:

https://github.com/dotnet/coreclr/blob/48ae85018b10f3d5f8ce340621cdac5ac6956066/src/jit/hwintrinsiclistxarch.h#L20

and this is used already for 10 SSE, SSE41 HW intrinsics.

2. Proposed implementation based on walking `GenTreeArgList` is several times slower from my implementation, code size is similar, and will be even more inefficient in comparison to my implementation when used to cover additional cases for 4, 8, 16, and 32 arguments which will be handled for `SetVector128` and `SetVector256` intrinsics. My implementation is O(1) and proposed implementation will be O(n) where n is number of arguments.

3. From API design perspective function should abstract only this functionality and parameters which are essential for its logic. `GenTreeHWIntrinsic* node` is not used for recovering number of arguments which depend on argument `var_type`. Therefore, signature which uses `Intrinsic ID` and `var_types` argument type is more universal and compact.

Reply:
> Proposed implementation based on walking GenTreeArgList is several times slower from my implementation,

That is why we should have the fast path:
```cpp
    if ((flagsOfHWIntrinsic(intrinsic) & HW_Flag_UnfixedNumArgs) == 0)
    {
        return hwIntrinsicInfoArray[intrinsic - NI_HW_INTRINSIC_START - 1].numArgs;
    }
```

> when used to cover additional cases for 4, 8, 16, and 32 arguments which will be handled for SetVector128 and SetVector256 intrinsics. 

We have talked so many times. ***These intrinsics are "NoCodeGen" helpers***, they only exist in the front-end (importer), and we can use `sig->numArgs` in the importer.

> My implementation is O(1) and proposed implementation will be O(n) where n is number of arguments.

They are both O(1), the "numArgs" only can be 0, 1, 2, or 3.

Reply:
>  GenTreeHWIntrinsic* node is not used for recovering number of arguments 

In the back-end, we already have the information in the IR, why not use it?

> which depend on argument var_type.

Most of the "UnfixedNumArgs" intrinsics cannot depend on the base type.

```csharp
public static Vector128<float> ReciprocalScalar(Vector128<float> value) { throw new PlatformNotSupportedException(); }
public static Vector128<float> ReciprocalScalar(Vector128<float> upper, Vector128<float> value) { throw new PlatformNotSupportedException(); }
```

Reply:
> It introduces additional flag HW_Flag_UnfixedNumArgs which is not necessary 

Yes, I just feel it more reliable, not strongly.

Reply:
> That is why we should have the fast path

There is a fast path in my implementation as well:

https://github.com/dotnetrt/coreclr/blob/10397525cb4befffdd33421c852de89064914eef/src/jit/hwintrinsicxarch.cpp#L252

> In the back-end, we already have the information in the IR, why not use it?

It is slower than accessing info from static `int Compiler::numArgsOfHWIntrinsic(NamedIntrinsic intrinsic, var_types baseType)`, the other overload is called anyway so replacing in 3 places call to `int Compiler::numArgsOfHWIntrinsic(NamedIntrinsic intrinsic)` with cal to new overload does not create additional cost.

> Most of the "UnfixedNumArgs" intrinsics cannot depend on the base type.

Yes, but they use different code path for emitting instruction which does not depend directly on correct value of `numArgs`.

> they only exist in the front-end (importer), and we can use sig->numArgs in the importer.

Yes, helpers are `abstract` since they do not represent specific CPU instructions, however, to efficiently implement larger ones it is easier to emit them in code gen instead of creating them in importer. Similar approach to emitting jump table for imm intrinsics and it directly depends on `numArgs`. Since `numArgs` are recovered every time at the beginning of `void CodeGen::genHWIntrinsic(GenTreeHWIntrinsic* node)` implementation:

https://github.com/dotnetrt/coreclr/blob/579a41cdf868d0cfba5bff071c782da0c59d1e0f/src/jit/hwintrinsiccodegenxarch.cpp#L59

we have this information already available.